src/libertem/api.py

```diff
@@ -559,7 +567,8 @@ def create_pick_analysis(self, dataset, x: int, y: int = None,
         parameters = {name: loc[name] for name in ['x', 'y', 'z'] if loc[name] is not None}
         return PickFrameAnalysis(dataset=dataset, parameters=parameters)
 
-    def run(self, job: Union[Job, Analysis], roi=None) -> Union[np.ndarray, AnalysisResultSet]:
+    def run(self, job: BaseAnalysis,
```

Comment:
Like you can see here, the type of job was `Union[Job, Analysis]` but `Job.TYPE` and `Analysis.TYPE` does not exists. After understanding the codebase I learned that `BaseAnalysis` does have `TYPE` in it but Analysis does not have TYPE. So I updated it with `BaseAnalysis`

Reply:
The problem with this approach is: the `run` function does need to accept objects of type `Job`, otherwise there would be no way to run them! The approach to fix a type error is not always so straight forward, in this case, the change needs to happen somewhere else.

Reply:
Then I think it should be better to leave this as `Union[Job, BaseAnalysis]`

Reply:
Actually, instead of `BaseAnalysis` it should be `Analysis`. I guess this gives an error because `Analysis` doesn't have the `TYPE` attribute, among other things? If so, this is actually a problem with the `BaseAanlysis` and `Analysis` classes - the `Analysis` class should define the complete interface (including attributes), and `BaseAnalysis` should only contain some default implementations.

This actually shows very well the value of these type annotations - they catch mistakes that slipped through the code review.

Let me know if you have further questions!

Reply:
I get it now, but I think if we implement `mypy` now in CI then this error will be there. Is there anyway that `mypy` passes?

Reply:
For the beginning, we can make a separate job in travis for mypy which is allowed to fail. That way, we get kind of a TODO list of errors to fix out of the CI job. If we type check the whole code base, we will have many error messages anyway.

Reply:
That makes sense. Then I will try to implement `mypy` too.